# Getting started

To use pandas, you'll typically start with the following line of code.

In [101]:
import pandas as pd

# Creating data

There are two core objects in pandas: the **DataFrame** and the **Series**.

### DataFrame

A DataFrame is a table. It contains an array of individual *entries*, each of which has a certain *value*. Each entry corresponds to a row (or *record*) and a *column*.

For example, consider the following simple DataFrame:

In [102]:
pd.DataFrame({'Yes': [50, 21], 'No': [131, 2]})

Yes   No
0   50  131
1   21    2

In this example, the "0, No" entry has the value of 131. The "0, Yes" entry has a value of 50, and so on.

DataFrame entries are not limited to integers. For instance, here's a DataFrame whose values are strings:

In [103]:
pd.DataFrame({'Bob': ['I liked it.', 'It was awful.'], 'Sue': ['Pretty good.', 'Bland.']})

Bob           Sue
0    I liked it.  Pretty good.
1  It was awful.        Bland.

We are using the `pd.DataFrame()` constructor to generate these DataFrame objects. The syntax for declaring a new one is a dictionary whose keys are the column names (`Bob` and `Sue` in this example), and whose values are a list of entries. This is the standard way of constructing a new DataFrame, and the one you are most likely to encounter.

The dictionary-list constructor assigns values to the *column labels*, but just uses an ascending count from 0 (0, 1, 2, 3, ...) for the *row labels*. Sometimes this is OK, but oftentimes we will want to assign these labels ourselves.

The list of row labels used in a DataFrame is known as an **Index**. We can assign values to it by using an `index` parameter in our constructor:

In [104]:
pd.DataFrame({'Bob': ['I liked it.', 'It was awful.'], 
              'Sue': ['Pretty good.', 'Bland.']},
             index=['Product A', 'Product B'])

Bob           Sue
Product A    I liked it.  Pretty good.
Product B  It was awful.        Bland.

### Series

A Series, by contrast, is a sequence of data values. If a DataFrame is a table, a Series is a list. And in fact you can create one with nothing more than a list:

In [105]:
pd.Series([1, 2, 3, 4, 5])

0    1
1    2
2    3
3    4
4    5
dtype: int64

A Series is, in essence, a single column of a DataFrame. So you can assign row labels to the Series the same way as before, using an `index` parameter. However, a Series does not have a column name, it only has one overall `name`:

In [106]:
pd.Series([30, 35, 40], index=['2015 Sales', '2016 Sales', '2017 Sales'], name='Product A')

2015 Sales    30
2016 Sales    35
2017 Sales    40
Name: Product A, dtype: int64

The Series and the DataFrame are intimately related. It's helpful to think of a DataFrame as actually being just a bunch of Series "glued together". We'll see more of this in the next section of this tutorial.

# Reading data files

Being able to create a DataFrame or Series by hand is handy. But, most of the time, we won't actually be creating our own data by hand. Instead, we'll be working with data that already exists.

Data can be stored in any of a number of different forms and formats. By far the most basic of these is the humble CSV file. When you open a CSV file you get something that looks like this:

```
Product A,Product B,Product C,
30,21,9,
35,34,1,
41,11,11
```

So a CSV file is a table of values separated by commas. Hence the name: "Comma-Separated Values", or CSV.

Let's now set aside our toy datasets and see what a real dataset looks like when we read it into a DataFrame. We'll use the `pd.read_csv()` function to read the data into a DataFrame. This goes thusly:

In [107]:
IMDB_data = pd.read_csv("../Datasets/IMDB-Movie-Data.csv")

We can use the `shape` attribute to check how large the resulting DataFrame is:

In [108]:
IMDB_data.shape

(1000, 12)

So our new DataFrame has 130,000 records split across 14 different columns. That's almost 2 million entries!

We can examine the contents of the resultant DataFrame using the `head()` command, which grabs the first five rows:

In [109]:
IMDB_data.head()

Rank                    Title                     Genre  \
0     1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1     2               Prometheus  Adventure,Mystery,Sci-Fi   
2     3                    Split           Horror,Thriller   
3     4                     Sing   Animation,Comedy,Family   
4     5            Suicide Squad  Action,Adventure,Fantasy   

                                         Description              Director  \
0  A group of intergalactic criminals are forced ...            James Gunn   
1  Following clues to the origin of mankind, a te...          Ridley Scott   
2  Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3  In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4  A secret government agency recruits some of th...            David Ayer   

                                              Actors  Year  Runtime (Minutes)  \
0  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014                121   
1  Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012                124   
2  James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016                117   
3  Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016                108   
4  Will Smith, Jared Leto, Margot Robbie, Viola D...  2016                123   

   Rating   Votes  Revenue (Millions)  Metascore  
0     8.1  757074              333.13       76.0  
1     7.0  485820              126.46       65.0  
2     7.3  157606              138.12       62.0  
3     7.2   60545              270.32       59.0  
4     6.2  393727              325.02       40.0

The `pd.read_csv()` function is well-endowed, with over 30 optional parameters you can specify. For example, you can see in this dataset that the CSV file has a built-in index, which pandas did not pick up on automatically. To make pandas use that column for the index (instead of creating a new one from scratch), we can specify an `index_col`.

In [110]:
IMDB_data = pd.read_csv("../Datasets/IMDB-Movie-Data.csv", index_col=0)
IMDB_data.head()

Title                     Genre  \
Rank                                                      
1     Guardians of the Galaxy   Action,Adventure,Sci-Fi   
2                  Prometheus  Adventure,Mystery,Sci-Fi   
3                       Split           Horror,Thriller   
4                        Sing   Animation,Comedy,Family   
5               Suicide Squad  Action,Adventure,Fantasy   

                                            Description              Director  \
Rank                                                                            
1     A group of intergalactic criminals are forced ...            James Gunn   
2     Following clues to the origin of mankind, a te...          Ridley Scott   
3     Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
4     In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
5     A secret government agency recruits some of th...            David Ayer   

                                                 Actors  Year  \
Rank                                                            
1     Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
2     Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
3     James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
4     Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
5     Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   

      Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
Rank                                                                    
1                   121     8.1  757074              333.13       76.0  
2                   124     7.0  485820              126.46       65.0  
3                   117     7.3  157606              138.12       62.0  
4                   108     7.2   60545              270.32       59.0  
5                   123     6.2  393727              325.02       40.0

So a CSV file is a table of values separated by commas. Hence the name: "Comma-Separated Values", or CSV.

Let's now set aside our toy datasets and see what a real dataset looks like when we read it into a DataFrame. We'll use the `pd.read_csv()` function to read the data into a DataFrame. This goes thusly:

# Indexing, Selecting & Assigning

## Native Accessors

Native Python objects provide  good ways of indexing data. Pandas carries all of these over, which helps make it easy to start with.

Consider this DataFrame:

In [111]:
IMDB_data = pd.read_csv("../Datasets/IMDB-Movie-Data.csv", index_col=0)
pd.set_option('display.max_rows', 5)
IMDB_data

Title                     Genre  \
Rank                                                      
1     Guardians of the Galaxy   Action,Adventure,Sci-Fi   
2                  Prometheus  Adventure,Mystery,Sci-Fi   
...                       ...                       ...   
999              Search Party          Adventure,Comedy   
1000               Nine Lives     Comedy,Family,Fantasy   

                                            Description          Director  \
Rank                                                                        
1     A group of intergalactic criminals are forced ...        James Gunn   
2     Following clues to the origin of mankind, a te...      Ridley Scott   
...                                                 ...               ...   
999   A pair of friends embark on a mission to reuni...    Scot Armstrong   
1000  A stuffy businessman finds himself trapped ins...  Barry Sonnenfeld   

                                                 Actors  Year  \
Rank                                                            
1     Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
2     Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
...                                                 ...   ...   
999   Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
1000  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

      Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
Rank                                                                    
1                   121     8.1  757074              333.13       76.0  
2                   124     7.0  485820              126.46       65.0  
...                 ...     ...     ...                 ...        ...  
999                  93     5.6    4881                 NaN       22.0  
1000                 87     5.3   12435               19.64       11.0  

[1000 rows x 11 columns]

In Python, we can access the property of an object by accessing it as an attribute. A `book` object, for example, might have a `title` property, which we can access by calling `book.title`. Columns in a pandas DataFrame work in much the same way. 

Hence to access the `title` property of `IMDB_data` we can use:

In [112]:
IMDB_data.Title

Rank
1       Guardians of the Galaxy
2                    Prometheus
                 ...           
999                Search Party
1000                 Nine Lives
Name: Title, Length: 1000, dtype: object

If we have a Python dictionary, we can access its values using the indexing (`[]`) operator. We can do the same with columns in a DataFrame:

In [113]:
IMDB_data['Title']

Rank
1       Guardians of the Galaxy
2                    Prometheus
                 ...           
999                Search Party
1000                 Nine Lives
Name: Title, Length: 1000, dtype: object

These are the two ways of selecting a specific Title out of a DataFrame. Neither of them is more or less syntactically valid than the other, but the indexing operator `[]` does have the advantage that it can handle column names with reserved characters in them (e.g. if we had a `Revenue (Millions)` column, `reviews.Revenue (Millions)` wouldn't work).

Doesn't a pandas Series look kind of like a fancy dictionary? It pretty much is, so it's no surprise that, to drill down to a single specific value, we need only use the indexing operator `[]` once more:

In [114]:
IMDB_data['Title'][1]

'Guardians of the Galaxy'

## Indexing in Pandas

The indexing operator and attribute selection are nice because they work just like they do in the rest of the Python ecosystem. As a novice, this makes them easy to pick up and use. However, pandas has its own accessor operators, `loc` and `iloc`. For more advanced operations, these are the ones you're supposed to be using.

#### Index-based selection

Pandas indexing works in one of two paradigms. The first is **index-based selection**: selecting data based on its numerical position in the data. `iloc` follows this paradigm.

To select the first row of data in a DataFrame, we may use the following:

In [115]:
IMDB_data.iloc[0]

Title                 Guardians of the Galaxy
Genre                 Action,Adventure,Sci-Fi
                               ...           
Revenue (Millions)                     333.13
Metascore                                76.0
Name: 1, Length: 11, dtype: object

In [116]:
IMDB_data.iloc[:, 0]

Rank
1       Guardians of the Galaxy
2                    Prometheus
                 ...           
999                Search Party
1000                 Nine Lives
Name: Title, Length: 1000, dtype: object

In [117]:
IMDB_data.iloc[:3, 0]

Rank
1    Guardians of the Galaxy
2                 Prometheus
3                      Split
Name: Title, dtype: object

In [118]:
IMDB_data.iloc[1:3, 0]

Rank
2    Prometheus
3         Split
Name: Title, dtype: object

In [119]:
IMDB_data.iloc[[0, 1, 2], 0]

Rank
1    Guardians of the Galaxy
2                 Prometheus
3                      Split
Name: Title, dtype: object

In [120]:
IMDB_data.iloc[-5:]

Title                  Genre  \
Rank                                                  
996     Secret in Their Eyes    Crime,Drama,Mystery   
997          Hostel: Part II                 Horror   
998   Step Up 2: The Streets    Drama,Music,Romance   
999             Search Party       Adventure,Comedy   
1000              Nine Lives  Comedy,Family,Fantasy   

                                            Description          Director  \
Rank                                                                        
996   A tight-knit team of rising investigators, alo...         Billy Ray   
997   Three American college students studying abroa...          Eli Roth   
998   Romantic sparks occur between two dance studen...        Jon M. Chu   
999   A pair of friends embark on a mission to reuni...    Scot Armstrong   
1000  A stuffy businessman finds himself trapped ins...  Barry Sonnenfeld   

                                                 Actors  Year  \
Rank                                                            
996   Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
997   Lauren German, Heather Matarazzo, Bijou Philli...  2007   
998   Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
999   Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
1000  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

      Runtime (Minutes)  Rating  Votes  Revenue (Millions)  Metascore  
Rank                                                                   
996                 111     6.2  27585                 NaN       45.0  
997                  94     5.5  73152               17.54       46.0  
998                  98     6.2  70699               58.01       50.0  
999                  93     5.6   4881                 NaN       22.0  
1000                 87     5.3  12435               19.64       11.0

#### Label-based Selection

The second paradigm for attribute selection is the one followed by the `loc` operator: **label-based selection**. In this paradigm, it's the data index value, not its position, which matters.

For example, to get the first entry in `reviews`, we would now do the following:

In [121]:
IMDB_data.loc[1, 'Actors']

'Chris Pratt, Vin Diesel, Bradley Cooper, Zoe Saldana'

`iloc` is conceptually simpler than `loc` because it ignores the dataset's indices. When we use `iloc` we treat the dataset like a big matrix (a list of lists), one that we have to index into by position. `loc`, by contrast, uses the information in the indices to do its work. Since your dataset usually has meaningful indices, it's usually easier to do things using `loc` instead. For example, here's one operation that's much easier using `loc`:

In [122]:
IMDB_data.loc[:, ['Genre','Director','Votes']]

Genre          Director   Votes
Rank                                                    
1      Action,Adventure,Sci-Fi        James Gunn  757074
2     Adventure,Mystery,Sci-Fi      Ridley Scott  485820
...                        ...               ...     ...
999           Adventure,Comedy    Scot Armstrong    4881
1000     Comedy,Family,Fantasy  Barry Sonnenfeld   12435

[1000 rows x 3 columns]

#### Choosing between `loc` and `iloc`

When choosing or transitioning between `loc` and `iloc`, there is one "gotcha" worth keeping in mind, which is that the two methods use slightly different indexing schemes.

`iloc` uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded. So `0:10` will select entries `0,...,9`. `loc`, meanwhile, indexes inclusively. So `0:10` will select entries `0,...,10`.

Why the change? Remember that loc can index any stdlib type: strings, for example. If we have a DataFrame with index values `Apples, ..., Potatoes, ...`, and we want to select "all the alphabetical fruit choices between Apples and Potatoes", then it's a lot more convenient to index `df.loc['Apples':'Potatoes']` than it is to index something like `df.loc['Apples', 'Potatoet']` (`t` coming after `s` in the alphabet).

This is particularly confusing when the DataFrame index is a simple numerical list, e.g. `0,...,1000`. In this case `df.iloc[0:1000]` will return 1000 entries, while `df.loc[0:1000]` return 1001 of them! To get 1000 elements using `loc`, you will need to go one lower and ask for `df.loc[0:999]`. 

Otherwise, the semantics of using `loc` are the same as those for `iloc`.

## Manipulating The Index

Label-based selection derives its power from the labels in the index. Critically, the index we use is not immutable. We can manipulate the index in any way we see fit.

The `set_index()` method can be used to do the job. Here is what happens when we `set_index` to the `title` field:

In [123]:
IMDB_data.set_index('Title')

Genre  \
Title                                               
Guardians of the Galaxy   Action,Adventure,Sci-Fi   
Prometheus               Adventure,Mystery,Sci-Fi   
...                                           ...   
Search Party                     Adventure,Comedy   
Nine Lives                  Comedy,Family,Fantasy   

                                                               Description  \
Title                                                                        
Guardians of the Galaxy  A group of intergalactic criminals are forced ...   
Prometheus               Following clues to the origin of mankind, a te...   
...                                                                    ...   
Search Party             A pair of friends embark on a mission to reuni...   
Nine Lives               A stuffy businessman finds himself trapped ins...   

                                 Director  \
Title                                       
Guardians of the Galaxy        James Gunn   
Prometheus                   Ridley Scott   
...                                   ...   
Search Party               Scot Armstrong   
Nine Lives               Barry Sonnenfeld   

                                                                    Actors  \
Title                                                                        
Guardians of the Galaxy  Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...   
Prometheus               Noomi Rapace, Logan Marshall-Green, Michael Fa...   
...                                                                    ...   
Search Party             Adam Pally, T.J. Miller, Thomas Middleditch,Sh...   
Nine Lives               Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...   

                         Year  Runtime (Minutes)  Rating   Votes  \
Title                                                              
Guardians of the Galaxy  2014                121     8.1  757074   
Prometheus               2012                124     7.0  485820   
...                       ...                ...     ...     ...   
Search Party             2014                 93     5.6    4881   
Nine Lives               2016                 87     5.3   12435   

                         Revenue (Millions)  Metascore  
Title                                                   
Guardians of the Galaxy              333.13       76.0  
Prometheus                           126.46       65.0  
...                                     ...        ...  
Search Party                            NaN       22.0  
Nine Lives                            19.64       11.0  

[1000 rows x 10 columns]

This is useful if you can come up with an index for the dataset which is better than the current one.

## Conditional Selection


So far, we’ve been indexing various aspects of the data, utilizing the structural properties of the DataFrame itself. To perform interesting analyses with the data, however, we often need to formulate questions based on specific conditions.

For example, suppose that we’re particularly interested in movies with a Rating higher than average.

We can begin by checking if each movie meets this condition:

In [124]:
IMDB_data.Rating > 7.5

Rank
1        True
2       False
        ...  
999     False
1000    False
Name: Rating, Length: 1000, dtype: bool

This operation produced a Series of `True`/`False` booleans based on the `Rating` of each record.  This result can then be used inside of `loc` to select the relevant data:

In [125]:
IMDB_data.loc[IMDB_data.Rating > 7.5]

Title                    Genre  \
Rank                                                     
1     Guardians of the Galaxy  Action,Adventure,Sci-Fi   
7                  La La Land       Comedy,Drama,Music   
...                       ...                      ...   
980        The Skin I Live In           Drama,Thriller   
992          Taare Zameen Par       Drama,Family,Music   

                                            Description         Director  \
Rank                                                                       
1     A group of intergalactic criminals are forced ...       James Gunn   
7     A jazz pianist falls for an aspiring actress i...  Damien Chazelle   
...                                                 ...              ...   
980   A brilliant plastic surgeon, haunted by past t...  Pedro Almodóvar   
992   An eight-year-old boy is thought to be a lazy ...       Aamir Khan   

                                                 Actors  Year  \
Rank                                                            
1     Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
7     Ryan Gosling, Emma Stone, Rosemarie DeWitt, J....  2016   
...                                                 ...   ...   
980   Antonio Banderas, Elena Anaya, Jan Cornet,Mari...  2011   
992   Darsheel Safary, Aamir Khan, Tanay Chheda, Sac...  2007   

      Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
Rank                                                                    
1                   121     8.1  757074              333.13       76.0  
7                   128     8.3  258682              151.06       93.0  
...                 ...     ...     ...                 ...        ...  
980                 120     7.6  108772                3.19       70.0  
992                 165     8.5  102697                1.20       42.0  

[195 rows x 11 columns]

This DataFrame has ~195 rows. The original had 1000 rows. This means that about 20% of them have a review score above 7.5.

We might want to review recently released movies.

We can use the ampersand (`&`) to combine two questions:

In [126]:
IMDB_data.loc[(IMDB_data.Rating > 7.5) & (IMDB_data.Year >= 2014)]

Title                       Genre  \
Rank                                                               
1            Guardians of the Galaxy     Action,Adventure,Sci-Fi   
7                         La La Land          Comedy,Drama,Music   
...                              ...                         ...   
904       How to Train Your Dragon 2  Animation,Action,Adventure   
951   Me and Earl and the Dying Girl                Comedy,Drama   

                                            Description             Director  \
Rank                                                                           
1     A group of intergalactic criminals are forced ...           James Gunn   
7     A jazz pianist falls for an aspiring actress i...      Damien Chazelle   
...                                                 ...                  ...   
904   When Hiccup and Toothless discover an ice cave...         Dean DeBlois   
951   High schooler Greg, who spends most of his tim...  Alfonso Gomez-Rejon   

                                                 Actors  Year  \
Rank                                                            
1     Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
7     Ryan Gosling, Emma Stone, Rosemarie DeWitt, J....  2016   
...                                                 ...   ...   
904   Jay Baruchel, Cate Blanchett, Gerard Butler, C...  2014   
951   Thomas Mann, RJ Cyler, Olivia Cooke, Nick Offe...  2015   

      Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
Rank                                                                    
1                   121     8.1  757074              333.13       76.0  
7                   128     8.3  258682              151.06       93.0  
...                 ...     ...     ...                 ...        ...  
904                 102     7.9  237565              177.00       76.0  
951                 105     7.8   92076                6.74       74.0  

[75 rows x 11 columns]

Let's say a movie with a rating above 7.5 or released after 2014 is under review. For this we use a pipe (`|`):

In [127]:
IMDB_data.loc[(IMDB_data.Rating > 7.5) | (IMDB_data.Year >= 2014)]

Title                    Genre  \
Rank                                                     
1     Guardians of the Galaxy  Action,Adventure,Sci-Fi   
3                       Split          Horror,Thriller   
...                       ...                      ...   
999              Search Party         Adventure,Comedy   
1000               Nine Lives    Comedy,Family,Fantasy   

                                            Description            Director  \
Rank                                                                          
1     A group of intergalactic criminals are forced ...          James Gunn   
3     Three girls are kidnapped by a man with a diag...  M. Night Shyamalan   
...                                                 ...                 ...   
999   A pair of friends embark on a mission to reuni...      Scot Armstrong   
1000  A stuffy businessman finds himself trapped ins...    Barry Sonnenfeld   

                                                 Actors  Year  \
Rank                                                            
1     Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
3     James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
...                                                 ...   ...   
999   Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
1000  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

      Runtime (Minutes)  Rating   Votes  Revenue (Millions)  Metascore  
Rank                                                                    
1                   121     8.1  757074              333.13       76.0  
3                   117     7.3  157606              138.12       62.0  
...                 ...     ...     ...                 ...        ...  
999                  93     5.6    4881                 NaN       22.0  
1000                 87     5.3   12435               19.64       11.0  

[642 rows x 11 columns]

Pandas comes with a few built-in conditional selectors, two of which we will highlight here. 

The first is `isin`. `isin` is lets you select data whose value "is in" a list of values.

In [128]:
IMDB_data.loc[IMDB_data.Genre.isin(['Horror'])]

Title   Genre  \
Rank                                    
43    Don't Fuck in the Woods  Horror   
259                Lights Out  Horror   
...                       ...     ...   
989                   Martyrs  Horror   
997           Hostel: Part II  Horror   

                                            Description           Director  \
Rank                                                                         
43    A group of friends are going on a camping trip...      Shawn Burkett   
259   Rebecca must unlock the terror behind her litt...  David F. Sandberg   
...                                                 ...                ...   
989   A young woman's quest for revenge against the ...     Pascal Laugier   
997   Three American college students studying abroa...           Eli Roth   

                                                 Actors  Year  \
Rank                                                            
43    Brittany Blanton, Ayse Howard, Roman Jossart,N...  2016   
259   Teresa Palmer, Gabriel Bateman, Maria Bello,Bi...  2016   
...                                                 ...   ...   
989   Morjana Alaoui, Mylène Jampanoï, Catherine Bég...  2008   
997   Lauren German, Heather Matarazzo, Bijou Philli...  2007   

      Runtime (Minutes)  Rating  Votes  Revenue (Millions)  Metascore  
Rank                                                                   
43                   73     2.7    496                 NaN        NaN  
259                  81     6.4  69823               67.24       58.0  
...                 ...     ...    ...                 ...        ...  
989                  99     7.1  63785                 NaN       89.0  
997                  94     5.5  73152               17.54       46.0  

[11 rows x 11 columns]

The second is `isnull` (and its companion `notnull`). These methods let you highlight values which are (or are not) empty (`NaN`).

In [129]:
IMDB_data.loc[IMDB_data['Revenue (Millions)'].isnull()]

Title                Genre  \
Rank                                              
8                 Mindhorn               Comedy   
23          Hounds of Love   Crime,Drama,Horror   
...                    ...                  ...   
996   Secret in Their Eyes  Crime,Drama,Mystery   
999           Search Party     Adventure,Comedy   

                                            Description        Director  \
Rank                                                                      
8     A has-been actor best known for playing the ti...      Sean Foley   
23    A cold-blooded predatory couple while cruising...       Ben Young   
...                                                 ...             ...   
996   A tight-knit team of rising investigators, alo...       Billy Ray   
999   A pair of friends embark on a mission to reuni...  Scot Armstrong   

                                                 Actors  Year  \
Rank                                                            
8     Essie Davis, Andrea Riseborough, Julian Barrat...  2016   
23    Emma Booth, Ashleigh Cummings, Stephen Curry,S...  2016   
...                                                 ...   ...   
996   Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
999   Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   

      Runtime (Minutes)  Rating  Votes  Revenue (Millions)  Metascore  
Rank                                                                   
8                    89     6.4   2490                 NaN       71.0  
23                  108     6.7   1115                 NaN       72.0  
...                 ...     ...    ...                 ...        ...  
996                 111     6.2  27585                 NaN       45.0  
999                  93     5.6   4881                 NaN       22.0  

[128 rows x 11 columns]

## Assigning Data

Going the other way, assigning data to a DataFrame is easy. You can assign either a constant value:

In [130]:
IMDB_data['Audience'] = 'everyone'
IMDB_data['Audience']

Rank
1       everyone
2       everyone
          ...   
999     everyone
1000    everyone
Name: Audience, Length: 1000, dtype: object

Or with an iterable of values:

In [131]:
IMDB_data['index_backwards'] = range(len(IMDB_data), 0, -1)
IMDB_data['index_backwards']

Rank
1       1000
2        999
        ... 
999        2
1000       1
Name: index_backwards, Length: 1000, dtype: int64